torchtitan/datasets/experimental_datasets.py

```diff
@@ -0,0 +1,1212 @@
+Cleanup is planned for the future. 
+"""
+
+
+# --------------  UTILITIES  --------------
+
+
+def _get_latest(targdir, qualifier=lambda x: True):
+    """Fetch the latest file or folder written to target directory, subject to name passing the qualifier fn.
+    If directory is empty or nonexistent or no items qualify, return None."""
+    if os.path.exists(targdir) and len(os.listdir(targdir)) > 0:
+        latest = max(
+            [
+                os.path.join(targdir, x)
+                for x in os.listdir(targdir)
+                if qualifier(os.path.join(targdir, x))
+            ],
+            key=lambda path: int(path.split("/")[-1].split("-")[1]),
+        )
+        return os.path.join(targdir, latest)
+    return None
+
+
+def _shard_partition(itemlist: List[Any], rank: int, worldsize: int) -> List[Any]:
+    """
+    Partition itemlist into worldsize chunks, grab chunk corresponding to rank and return.
+    """
+    return itemlist[
+        (rank * len(itemlist)) // worldsize : ((rank + 1) * len(itemlist)) // worldsize
+    ]
+
+
+def _shard_inclusive(itemlist: List[Any], rank: int, worldsize: int) -> List[Any]:
+    """
+    In cases where len(itemlist) % worldsize != 0, allow for fractional ownership of items,
+    and return the span including all owned items, fractional or otherwise.
+    """
+    start = math.floor(len(itemlist) * rank / worldsize)
+    end = math.ceil(len(itemlist) * (rank + 1) / worldsize)
+    return itemlist[start:end]
+
+
+class _Stateful_Dataset(data.IterableDataset):
+    """
+    Stub for stateful datasets, extends data.IterableDataset with state_dict methods.
+    All subclasses should specify the params to be considered stateful or reshardable in the
+    self.state_params and self.reshard_params lists.
+    """
+
+    def __init__(
+        self,
+        rank: int,
+        worldsize: int,
+    ):
+        assert rank >= 0, f"Rank {rank} must be a positive integer"
+        assert (
+            worldsize > rank
+        ), f"Worldsize {worldsize} must be greater than rank {rank}"
+        self.state_params: List[str] = []
+        self.reshard_params: List[str] = []
+        self.rank = rank
+        self.worldsize = worldsize
+        self.load_worldsize = (
+            worldsize  # Enable calling load_state_dict() directly, assume no rescaling
+        )
+
+    def statename(self, x: str):
+        # Note that this naming convention implicitly disallows repeated layers in the dataset pipeline
+        return self.__class__.__name__ + "." + x
+
+    def state_dict(self):
+        """
+        Retrieve all state and reshard flags (each worker/process saves its own state dict shard)
+        """
+        return {
+            self.statename(flag): getattr(self, flag)
+            for flag in self.state_params + self.reshard_params
+        }
+
+    def _reshard(self, sharded_list):
+        """
+        Sharded_list is a list of lists, where each "shard" sublist must have the same length.
```

Comment:
I suppose it's not strictly necessary that shards have the same length, we just take that shortcut because this function is only ever used to reassign per-worker data structures from `load_worldsize` to `worldsize`. So far we've been able to assume that workers all hold the same fundamental data structures so this hasn't been an issue. This doesn't get called on any actual data shard files - apologies for unclear naming/commenting here